src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,682 @@
+/-
+Copyright (c) 2021 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import measure_theory.vector_measure
+import order.symm_diff
+
+/-!
+# Hahn decomposition
+
+This file defines positive and negative sets with respect to a vector measure and
+prove the Hahn decomposition theorem (signed version).
+
+## Main definitions
+
+* `vector_measure.positive` : a set `i` is positive with respect to a vector measure `v` if `v`
+  restricted on `i` is non-negative, i.e. `0 ≤ v.restrict i`.
+* `vector_measure.negative` : a set `i` is negative with respect to a vector measure `v` if `v`
+  restricted on `i` is non-positive, i.e. `v.restrict i ≤ 0`.
+
+## Main results
+
+* `signed_measure.exists_disjoint_positive_negative_union_eq` : the Hahn decomposition theorem.
+* `signed_measure.exists_negative_set` : A measurable set of negative measure contains at least
+  one negative subset.
+
+## Tags
+
+positive, negative, Hahn decomposition theorem
+-/
+
+noncomputable theory
+open_locale classical big_operators nnreal ennreal
+
+section lemmas
+
+variables {α : Type*}
+
+open set filter
+
+lemma exists_tendsto_Inf {S : set ℝ} (hS : ∃ x, x ∈ S) (hS' : ∃ x, ∀ y ∈ S, x ≤ y) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (nhds (Inf S)) :=
+begin
+  have : ∀ n : ℕ, ∃ t ∈ S, t < Inf S + 1 / (n + 1 : ℝ),
+  { exact λ n, (real.Inf_lt _ hS hS').1 ((lt_add_iff_pos_right _).2 nat.one_div_pos_of_nat) },
+  choose f hf using this,
+  refine ⟨f, λ n, (hf n).1, _⟩,
+  rw tendsto_iff_dist_tendsto_zero,
+  refine squeeze_zero_norm _ tendsto_one_div_add_at_top_nhds_0_nat,
+  intro n,
+  obtain ⟨hf₁, hf₂⟩ := hf n,
+  rw [real.dist_eq, real.norm_eq_abs, abs_abs,
+      abs_of_nonneg (sub_nonneg.2 (real.Inf_le S hS' hf₁))],
+  linarith,
+end
+
+lemma exists_tendsto_Sup {S : set ℝ} (hS : ∃ x, x ∈ S) (hS' : ∃ x, ∀ y ∈ S, y ≤ x) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (nhds (Sup S)) :=
+begin
```

Comment:
You shouldn't need to give a proof here: this should follow from `exists_tendsto_Inf` applied to `order_dual alpha`.